"Where to start" doc update (#242)
fredrikhgrelland/vagrant-hashistack: template/README.md

Step-by-step:
1. Create a [Dockerfile](https://docs.docker.com/engine/reference/builder/) and put it in `docker/`.
1. Use the box's startup playbooks to build and make the image available within the box. [Startup scheme](#default-startup-scheme). [Example building image](test_example/dev/ansible/01_build_docker_image.yml).
1. Create a [Nomad job specification](https://www.nomadproject.io/docs/job-specification). This should be placed in [conf/nomad/](conf/nomad/README.md). [Example hcl-file](test_example/conf/nomad/countdash.hcl)
1. Use terraform to deploy this job to Nomad. Refer to ["Terraform module requirements"](#terraform-module-requirements) on how to do this. [Example](test_example/example)

## File structure
|-- conf/						
|     |-- nomad/			# nomad job specifications
|   
|-- dev/             			# files used to control testing and development
|   |-- ansible/			# playbooks run at startup
|       |-- vagrant/			
|       |-- conf/		        # files altering the vagrant-box
|-- docker/				# files used to build docker-image
|
|-- example/				# terraform-files using the terraform-module
|

> |-- test_example/			# a copy of template, filled with example-code

Reviewer: This does not look well formatted. I tried the wrap it around triple quotes ``` and that did the job.
\```
|-- conf/
|     |-- nomad/			# nomad job specifications
|  
|-- dev/             			# files used to control testing and development
|   |-- ansible/			# playbooks run at startup
|       |-- vagrant/		
|       |-- conf/		        # files altering the vagrant-box
|-- docker/				# files used to build docker-image
|
|-- example/				# terraform-files using the terraform-module
|
|-- test_example/			# a copy of template, filled with example-code
\```

Author: Didn't even cross my mind. Thank you! :grinning: Fixed now

## Using the template and box

### Commands
There are several commands that help to run the vagrant-box:
- `make up` provisions a [vagrant-hashistack](https://github.com/fredrikhgrelland/vagrant-hashistack/) box on your machine. After the machine and hashistack are set up it will run the default startup scheme.

- `make clean` take down the provisioned box if there is any.

- `make update` downloads the newest version of the [vagrant-hashistack box](https://github.com/fredrikhgrelland/vagrant-hashistack/) from [vagrantcloud](https://vagrantcloud.com/fredrikhgrelland/hashistack).

- `make example` runs the the example in [test_example/](test_example/)

For full info, check `template/Makefile`

`!Note` - Makefile commands are not idempotent in the context of vagrant-box.  You could face the error of port collisions. Most of the cases it could happen because of the vagrant box has already been running.

> Once vagrant-box is running, you can use other options to iterate of deployment in the development stage. 

Reviewer: _iterate of deployment_??

Author: Typo, fixed!

Documentation on [parent repository](https://github.com/fredrikhgrelland/vagrant-hashistack#usage).

> ## Goals of this template

Reviewer 2: ```suggestion
## Using this template
```

> This template aims to supply a developer with the tools necessary to do three things:

Reviewer 2: ```suggestion
This template provides a developer with the tools necessary to do three things:
```

- Build a docker image

> - Build a terraform module

Author: Yes. Would you like adding that to the goals?

> - Build a terraform module

Reviewer 3: yes

> - Build a terraform module

Reviewer 3: yes

> - Build a terraform module

Author: I would argue that goes under `Build a terraform module` and `Test and develop`, what do you think?

> - Build a terraform module

Reviewer 3: let's leave it as it is. My comment was subjective, let us wait for user feedback.

- Test and develop these two within a hashistack ecosystem

To achieve this we utilise a vagrant-box built from [fredrikhgrelland/vagrant-hashistack](https://github.com/fredrikhgrelland/vagrant-hashistack/).

## Where do I start?
First of all I would recommend you read the section ["Using the template and box"](using-the-template-and-box) for a detailed look at the box's features and configurations. Below is a rough step-by-step going from a fresh template to a functioning terraform module. You can find example-code of everything below in [test_example/](test_example/).

Step-by-step:
1. Create a [Dockerfile](https://docs.docker.com/engine/reference/builder/) and put it in `docker/`.

> 1. Use the box's startup playbooks to build and make the image available within the box. [Startup scheme](default-startup-scheme). [Example building image](test_example/dev/ansible/01_build_docker_image.yml).

Reviewer 3: What is `(default-startup-scheme)`?

Author: that is a link to the Markdown title `## Defaul startup scheme`

Reviewer 3: does it work? I thought smth like (#default-startup-scheme)

Author: This has been fixed, but you were right, it was written wrong!

1. Create a [Nomad job specification](https://www.nomadproject.io/docs/job-specification). This should be placed in [conf/nomad/](conf/nomad/README.md). [Example hcl-file](test_example/conf/nomad/countdash.hcl)

> 1. Use terraform to deploy this job to Nomad. Refer to ["Terraform module requirements"](terraform-module-requirements) on how to do this.

Reviewer 3: Is it `(terraform-module-requirements)` mock right now?

Author: Are you asking if the `Terraform module requirements` section is a mock?

Reviewer 3: Question is related to other question above

Author: Right, as above: written wrong, been fixed

> 1. Use terraform to deploy this job to Nomad. Refer to ["Terraform module requirements"](terraform-module-requirements) on how to do this.

Reviewer 2: ```suggestion
1. Use terraform to deploy this job to Nomad. Refer to ["Terraform module requirements"](terraform-module-requirements) on how to do this. [Example](test_example/example)
```

> ## File structure
> - [conf/nomad/](conf/nomad/), nomad job specifications
> - [dev/](dev/), files used to control testing and development
>   - [dev/ansible](dev/ansible), playbooks run at startup
>   - [dev/vagrant/conf/](dev/vagrant/conf), files altering the vagrant-box  
> - [docker/](docker/), files used to build docker-image
> - [example/](example/), terraform-files using the terraform-module, [ref](terraform-module-requirements)
> - [test_example/](test_example/), a copy of template, filled with example-code
> 

Author: ```suggestion
```
After adding it I'm conflicted, I feel this is overkill. Comments before I remove it?

Reviewer 3: ```
|-- conf/						
|     |-- nomad/			# nomad job specifications
|   
|-- dev/				# files used to control testing and development
|     |-- ansible/			# playbooks run at startup
|	  |-- vagrant/			
|            |-- conf/		        # files altering the vagrant-box
|-- docker/				# files used to build docker-image
|
|-- example/				# terraform-files using the terraform-module
|
|-- test_example/			# a copy of template, filled with example-code

```

Author: Hm, this does not format nicely when I use it in the code, am I doing something wrong?

Reviewer 3: Probably you need to adjust, my previous comments is a draft.

Author: Neha fixed this. https://github.com/fredrikhgrelland/vagrant-hashistack/pull/242#discussion_r473919647
Now fixed!

## Terraform module requirements
[Hashiscorp documentation](https://www.terraform.io/docs/modules/index.html)

 This template follows Hashicorp's standards for terraform modules. In short the root should contain a `main.tf`, `variables.tf` and `outputs.tf`. These three form a module. In addition to this there should be an `example/` folder that uses the module. To get a feel for how to use this effectively in our case you can look at the files [main.tf](test_example/main.tf), [variables.tf](test_example/variables.tf), and [outputs.tf](test_example/outputs.tf) from [test_example/](test_example/).

## Using the template and box

> 

Reviewer 3: Draft, I would suggest moving `### Commands` section under here.
```suggestion

### Commands
There are several commands that help to run the vagrant-box:
- `make up` provisions a [vagrant-hashistack](https://github.com/fredrikhgrelland/vagrant-hashistack/) box on your machine. After the machine and hashistack are set up it will run the default startup scheme.

- `make clean` take down the provisioned box if there is any.

- `make update` downloads the newest version of the [vagrant-hashistack box](https://github.com/fredrikhgrelland/vagrant-hashistack/) from [vagrantcloud](https://vagrantcloud.com/fredrikhgrelland/hashistack).

- `make example` runs the the example in [test_example/](test_example/)

For full info, check `template/Makefile`

`!Note` - Makefile commands are not idempotent in the context of vagrant-box.  You could face the error of port collisions. Most of the cases it could happen because of the vagrant box has already been running.

Once vagrant-box is running, you can use other options to iterate of deployment in the development stage. 
```

Reviewer 3: last sentence ref to #232

Author: Yeah, these are really good additions

### Building and testing docker image
See docker [README.md](docker/README.md).

### Default startup scheme
The default startup scheme will first start the hashistack and MinIO, then run [playbook.yml](dev/ansible/playbook.yml), which in turn will run all ansible-playbooks inside [dev/ansible/](dev/ansible). For more details go [here](dev/ansible/README.md). The code that runs the first playbook can be found in [Vagrantfile](Vagrantfile). Although this is the default we have supplied you with a plethora of options to customize and alter these processes. Below is a detailed description of the __whole__ startup procedure and details.

> #### Detailed startup procedure

Reviewer 3: ```suggestion
#### Detailed startup procedure (vagrant-box lifecycle)
```

Author: That section does not exist anymore?

__User provided__ - Provided by the user to alter the box or template in some way

1. `/home/vagrant/.env_default` - _preloaded_ - default variables
1. `vagrant/.env` - _user provided_ - variables override [details](pre-packaged-configuration-switches)
1. `vagrant/.env_override` - _system provided_ - variables are overridden for test purposes
1. `vagrant/dev/vagrant/conf/pre_ansible.sh` - _user provided_ - script running before ansible bootstrap procedure, [details](dev/vagrant/conf/pre_bootstrap/README.md)
1. `vagrant/dev/vagrant/conf/pre_bootstrap/*.yml` - _user provided_ - pre bootstrap tasks, running before hashistack software runs and ready, [details](dev/vagrant/conf/README.md)
1. `/etc/ansible/bootstrap.yml` - _preloaded_ - verify ansible variables and software configuration, run hashistack software and MinIO & verify that it started correctly
1. `vagrant/conf/post_bootstrap/*.yml` - _user provided_ - poststart scripts, running after hasistack software runs and ready, [details](dev/vagrant/conf/pre_bootstrap/README.md)
1. `vagrant/dev/conf/pre_ansible.sh` - _user provided_ - script running after ansible bootstrap procedure, [details](dev/vagrant/conf/README.md)

> 1. `vagrant/ansible/*.yml` - _user provided_ - ansible tasks included in playbook, [details](pre-and-post-hashistack-procedure)

Reviewer: The linked text *details* does not seem to be working.

Author: Oops, thank you. Fixed now!

### Accessing Consul, Vault, Nomad, and MinIO
The default box will start Nomad, Vault, Consul, and MinIO bound on loopback and advertising on the ip `10.0.3.10`, which should be available on your local machine.
Portforwarding for nomad on port `4646` should bind to `127.0.0.1` and should allow you to use the nomad binary to post jobs directly. Consul and Vault has also been portforwarded, and are also available on `127.0.0.1` on port `8500` and `8200` respectively.
Minio is started on port `9000` and sharing /vagrant (your repo) from within the vagrant box.
- Nomad ui is available on [http://10.0.3.10:4646](http://10.0.3.10:4646) and all links to services should work.
- Consul ui is available on [http://10.0.3.10:8500](http://10.0.3.10:8500)
- Vault ui is available on [http://10.0.3.10:8200](http://10.0.3.10:8200)
- minio ui is available on [http://10.0.3.10:9000](http://10.0.3.10:9000) (minioadmin:minioadmin)

> ### Commands
> - `make up` provisions a [vagrant-hashistack](https://github.com/fredrikhgrelland/vagrant-hashistack/) box on your machine. After the machine and hashistack is set up it will run default startup scheme.
> 
> - `make clean` takes down the provisioned box if there is any.
> 
> - `make update` downloads the newest version of the [vagrant-hashistack box](https://github.com/fredrikhgrelland/vagrant-hashistack/) from [vagrantcloud](https://vagrantcloud.com/fredrikhgrelland/hashistack).
> 
> - `make example` runs the the example in [test_example/](test_example/)
> 

Reviewer 3: Move section under `## Using the template and box`
```suggestion

```

|   |-- main.tf
|   |-- variables.tf
|   |-- outputs.tf
```

## Using the template and box

### Commands
There are several commands that help to run the vagrant-box:
- `make install` installs all prerequisites

- `make up` provisions a [vagrant-hashistack](https://github.com/fredrikhgrelland/vagrant-hashistack/) box on your machine. After the machine and hashistack are set up it will run the default [startup scheme](#default-startup-scheme).

- `make clean` take down the provisioned box if there is any.

- `make update` downloads the newest version of the [vagrant-hashistack box](https://github.com/fredrikhgrelland/vagrant-hashistack/) from [vagrantcloud](https://vagrantcloud.com/fredrikhgrelland/hashistack).

- `make example` runs the example in [test_example/](test_example/)

For full info, check `template/Makefile`

`!Note` - Makefile commands are not idempotent in the context of vagrant-box.  You could face the error of port collisions. Most of the cases it could happen because of the vagrant box has already been running.

Once vagrant-box is running, you can use other options like the Nomad- and Terraform-CLIs to iterate over the deployment in the development stage.

> 

Reviewer 3: ```suggestion
### Iterration of development process
Once you start the box with one of the command `make dev`, `make up` or `make example`, 
you need simple way how to continiously deploy development changes.

There are several options.

1. You can install Hashicorp binaries on local machine, such as terraform and nomad.
Than you can deploy chnages to the vagrant-box using these binaries.

Example terraform:
\```
# From the local machine
terraform init
terraform apply
\```

Example nomad:
\```
nomad job run countdash.hcl
\```

2. Box instance has all binaries are installed and available in the PATH.
You can use ´vagrant ssh´ to call commands

\´´´
# remote 
vagrant ssh default -c 'cd /vagrant; terraform init; terraform apply'

# local
vagrant ssh default
cd /vagrant
terraform init
terraform apply
\´´´

Note: default is the name of running VM. You could also use VM `id`.
To get vm `id` check `vagrant global-status`.
```

Reviewer 3: Just remove `\` for 99, 103, 106, 108, 113, 122

Reviewer 3: We migrate this to another PR. RESOLVED for now
